### PR TITLE
Dev beta

### PR DESCRIPTION
**Pull request by i-Alex**

### CODE REVIEW

File: examples/simpleapp/mc_sc_workflow_example.md [hunk 1]
```diff
@@ -15,7 +15,7 @@ Build SDK components by using command (in the root of the SDK folder):
 
 Run Bootstrapping tool using command: 
 
-`java -jar tools/sctool/target/Sidechains-SDK-ScBootstrappingTools-0.2.0.jar`
+`java -jar tools/sctool/target/Sidechains-SDK-ScBootstrappingTools-0.2.0-SNAPSHOT.jar`
```
Comment (albertog78) [inline examples/simpleapp/mc_sc_workflow_example.md:18]: Automatize version replacing

File: pom.xml [hunk 2]
```diff
@@ -2,7 +2,7 @@
 	<modelVersion>4.0.0</modelVersion>
 	<groupId>com.horizen</groupId>
 	<artifactId>Sidechains</artifactId>
-	<version>0.2.0</version>
+	<version>0.2.0-SNAPSHOT</version>
```
Comment (albertog78) [inline pom.xml:5]: this should be 0.2.1-SNAPSHOT

File: qa/SidechainTestFramework/sc_test_framework.py [hunk 3]
```diff
@@ -111,7 +111,7 @@ def main(self):
                           help="Don't stop bitcoinds after the test execution")
         parser.add_option("--zendir", dest="zendir", default="ZenCore/src",
                           help="Source directory containing zend/zen-cli (default: %default)")
-        parser.add_option("--scjarpath", dest="scjarpath", default="../examples/simpleapp/target/Sidechains-SDK-simpleapp-0.2.0.jar;../examples/simpleapp/target/lib/* com.horizen.examples.SimpleApp", #New option. Main class path won't be needed in future
+        parser.add_option("--scjarpath", dest="scjarpath", default="../examples/simpleapp/target/Sidechains-SDK-simpleapp-0.2.0-SNAPSHOT.jar;../examples/simpleapp/target/lib/* com.horizen.examples.SimpleApp", #New option. Main class path won't be needed in future
```
Comment (albertog78) [inline qa/SidechainTestFramework/sc_test_framework.py:114]: version name changing should be automatized or retrieved at runtime

File: sdk/pom.xml [hunk 4]
```diff
@@ -2,7 +2,7 @@
 	<modelVersion>4.0.0</modelVersion>
 	<groupId>com.horizen</groupId>
 	<artifactId>Sidechains-SDK</artifactId>
-	<version>0.2.0</version>
+	<version>0.2.0-SNAPSHOT</version>
```
Comment (albertog78) [inline sdk/pom.xml:5]: this should be 0.2.1-SNAPSHOT

File: examples/simpleapp/pom.xml [hunk 5]
```diff
@@ -2,7 +2,7 @@
 	<modelVersion>4.0.0</modelVersion>
 	<groupId>com.horizen</groupId>
 	<artifactId>Sidechains-SDK-simpleapp</artifactId>
-	<version>0.2.0</version>
+	<version>0.2.0-SNAPSHOT</version>
```
Comment (i-Alex) [inline examples/simpleapp/pom.xml:5]: update the version + 0.1 to master/dev branches.

File: examples/simpleapp/src/main/java/com/horizen/examples/SimpleAppModule.java [hunk 6]
```diff
@@ -100,27 +101,28 @@ protected void configure() {
                 .annotatedWith(Names.named("ApplicationState"))
                 .toInstance(defaultApplicationState);
 
+        int storageMaxRollbackDepth = 720 * 2 + 1; //two consensus epochs length + 1
         bind(Storage.class)
                 .annotatedWith(Names.named("SecretStorage"))
-                .toInstance(IODBStorageUtil.getStorage(secretStore));
+                .toInstance(new VersionedLevelDbStorageAdapter(secretStore, storageMaxRollbackDepth));
```
Comment (i-Alex) [inline examples/simpleapp/src/main/java/com/horizen/examples/SimpleAppModule.java:107]: At the moment we don't allow sidechain developers to customize consensus epoch length.
Let's make `storageMaxRollbackDepth` as an internal hardcoded parameter of  `VersionedLevelDbStorageAdapter` class.
As for me, developers will be confused about specification of this argument on their side and can do a mistake by passing the value less then `720*2+1`.

File: sdk/src/main/scala/com/horizen/storage/leveldb/VersionedLevelDbStorageAdapter.scala [hunk 7]
```diff
@@ -48,25 +54,41 @@ class VersionedLevelDbStorageAdapter(pathToDB: String, keepVersions: Int) extend
...
+
+    require(!versionIsExistInStorage(version), "Version is already exist in storage")
+
     val convertedToUpdate = toUpdateAsScala.map(pair => (pair.getKey.data, pair.getValue.data))
     val convertedToRemove = toRemoveAsScala.map(_.data)
     dataBase.update(convertedToUpdate, convertedToRemove)(version)
   }
 
-  override def rollback(versionID: ByteArrayWrapper): Unit = dataBase.rollbackTo(versionID)
+  def versionIsExistInStorage(versionForSearch: ByteArrayWrapper): Boolean = {
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/storage/leveldb/VersionedLevelDbStorageAdapter.scala:68]: maybe make it `private` and rename to `isVersionExistInStorage` or even `isVersionExist`

File: sdk/src/main/scala/com/horizen/storage/leveldb/VersionedLevelDbStorageAdapter.scala [hunk 8]
```diff
@@ -48,25 +54,41 @@ class VersionedLevelDbStorageAdapter(pathToDB: String, keepVersions: Int) extend
...
+  }
 
   override def rollbackVersions(): JList[ByteArrayWrapper] = dataBase.versions.map(byteArrayToWrapper).asJava
 
   override def close(): Unit = dataBase.close()
 
-  def createDb(path: String): VersionedLDBKVStore = {
-    val dir = new File(path)
-    dir.mkdirs()
+  def createDb(path: File): VersionedLDBKVStore = {
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/storage/leveldb/VersionedLevelDbStorageAdapter.scala:85]: should be private

File: examples/simpleapp/pom.xml (hunk 5 above)
Comment (oaleksandrov-ip) [reply examples/simpleapp/pom.xml:5]: done

File: examples/simpleapp/src/main/java/com/horizen/examples/SimpleAppModule.java (hunk 6 above)
Comment (oaleksandrov-ip) [reply examples/simpleapp/src/main/java/com/horizen/examples/SimpleAppModule.java:107]: done

File: sdk/src/main/scala/com/horizen/storage/leveldb/VersionedLevelDbStorageAdapter.scala (hunk 7 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/storage/leveldb/VersionedLevelDbStorageAdapter.scala:68]: done

File: sdk/src/main/scala/com/horizen/storage/leveldb/VersionedLevelDbStorageAdapter.scala (hunk 8 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/storage/leveldb/VersionedLevelDbStorageAdapter.scala:85]: done

**Comment by oaleksandrov-ip** [comment]
no longer actual, not merged changes are moved to https://github.com/HorizenOfficial/Sidechains-SDK/pull/210

**Pull request closed without merging**